students/johnn/session09/mailroom.py

```diff
@@ -0,0 +1,688 @@
+    def last_name(self):
+        """ return the last name """
+        return self._last_name
+
+    @last_name.setter
+    def last_name(self, value):
+        """ set the last name """
+        self._last_name = value.title()
+
+    @property
+    def suffix(self):
+        """ return the suffix """
+        return self._suffix
+
+    @suffix.setter
+    def suffix(self, value):
+        """ set the suffix """
+        # special case roman numbers to all caps
+        if value.lower() in ["i", "ii", "iii", "iv", "v"]:
+            self._suffix=value.upper()
+        else:
+            self._suffix=value.title()
+
+    @property
+    def informal_name(self):
+        """ return full name minus the suffix """
+        return " ".join(filter(None, [self.first_name, self.middle_name, self.last_name]))
+
+    @property
+    def full_name(self):
+        """ return the formal, full name """
+        # if suffix, add with a comma
+        if self.suffix != "":
+            suffix=", " + self.suffix
+        else:
+            suffix=""    
+        return " ".join(filter(None, [self.first_name, self.middle_name, self.last_name])) + suffix
+
+    @full_name.setter
+    def full_name(self, full_name):
+
+        """ allow the donor information to be updated via full_name as well """
+
+        # for simplicity's sake, we make the assumption a suffix will follow a comma
+
+        # capture suffix, if present
+        suffix=""
+        if full_name.find(","):
+            try:
+                suffix=full_name.split(",")[1].strip()
+            except:
+                pass
+
+        # name with suffix removed
+        informal_name=full_name.split(",")[0].strip()
+
+        # we assume the last name is one word minus the suffix
+        last_name=informal_name.split()[-1]
+       
+        # build a list with everything to the left of the last name
+        everything_but_last=informal_name.split()[0:-1]
+
+        # pull the first name off
+        try:
+            first_name=everything_but_last.pop(0)
+        except IndexError:
+            # if we get an error, they probably gave us a malformed name
+            first_name=""
+
+        try:
+            # pull the middle, if exists
+            middle_name=everything_but_last.pop(0)
+        except IndexError:
+            middle_name=""
+
+        self.first_name = first_name
+        self.middle_name = middle_name
+        self.last_name = last_name
+        self.suffix = suffix
+
+    def _query_attributes(self, which_attributes=[], return_empty=True):
+        """ return list of formatted attribute/value entries """
+        attributes = []
+        for attr in which_attributes:
+            # for our purposes, never print internal attributes
+            if not attr.startswith("_"):
+                try:
+                    str_val=repr(getattr(self,attr))
+                    # don't return empty values if they don't want them
+                    if eval(str_val) or return_empty:
```

Comment:
I think this could be better done with some metaprogramming as well. So bring it up next quarter.

but to start -- a class instance  has a __dict__ attribute with all it's attributes in it -- you could probably make use of that.

Though I'm still not sure what you're trying to accomplish here...